Add sensor platform for vicare integration (heatpump)

crazyfx1 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
Removed existing attributes from the climate entity which are now available as separate sensors. 
If you were using one of the following attributes, you must migrate to the new sensors. The sensors are created automatically.

For all heating types:
* `supply_temperature`
* `outside_temperature`
* `circulationpump_active`

For heating type `gas`:
* `burner_modulation`
* `boiler_temperature`

For heating type: `heatpump`:
* `return_temperature`

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add sensor platform for vicare. The vicare api provides additional sensor data which are currently available in the attributes list of the device. This change moves the attributes out to separate sensors and enables even more sensor data (ie. consumption data for gas heatings) to be added in the future.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [X] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/13054

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

crazyfx1 committed:
Merge pull request #1 from home-assistant/dev

Merge from base

crazyfx1 committed:
Merge pull request #2 from home-assistant/dev

Update from base

crazyfx1 committed:
Add sensor platform for vicare (heatpump)

crazyfx1 committed:
Formatting and fixes

crazyfx1 committed:
Formatting and fixes 2

crazyfx1 committed:
Fixes and formatting 3

crazyfx1 committed:
Fixes and formatting 4

crazyfx1 commented:
@oischinger Maybe you can check this PR anytime soon? It might be also relevant for the other open PR to implement as sensors #32304 .

oischinger commented:
Thanks for pinging me. I forgot the about that one.

In general I guess it makes sense to also create sensors. I just wonder what would be the best general rule:
Should we create sensors for all attributes that are not really covered by the climate platform?
Should we create sensors for the additional temperature measurements as in this pr?
Should we create sensors for the consumption data (in watts or kWh)?

Maybe @MartinHjelmare can give us some pointers on what is the preferred strategy of home assistant for additional sensors of the climate platform?

MartinHjelmare commented:
Please create sensor entities for all measurements that are not directly covered by the base climate entity design.

oischinger commented:
Ok. I'll do some testing later today and then we can go ahead with this PR.

oischinger committed:
Add binary_sensor and more sensors

This moves some more climate attributes to sensors and adds
binary_sensors

oischinger commented:
Hi @crazyfx1 
In general the changes are fine but I believe it would be good to move all the relevant attributes to sensors in a single PR. I just uploaded a commit based on your PR here:
https://github.com/home-assistant/core/commit/dac5a3a050c7e08dd33220011955bfd22af5f415

Maybe you can add me as a contributor to your fork of the HA repositories then I can directly contribute to that commit this PR?

crazyfx1 commented:
I've added you as contributor.
I hope it's fine to add binary_sensor and sensor in one PR. I've read somewhere that you should only add one platform per PR but I can't find the rule anymore ;)

oischinger commented:
let's see what we get away with @crazyfx1  ;)

I added now also the consumption stuff and the missing burner attributes from this (stale?) PR:
https://github.com/home-assistant/core/pull/32304

Note that I also override the is_available function now to make entities unavailable when there is no data or the API fails for some reason.

oischinger reviewed (approved):
To me the changes look good now. I tested it with my gas heater. Would be good to have another check with a heatpump device before this goes upstream.

crazyfx1 commented:
@oischinger 
Thanks for the additional changes.

I think the binary_sensor for the problem is not correctly implemented for heatpumps. 
What's your result for getActiveError() of PyViCare?
Mine is:
```JSON
{"new": [], "current": [], "gone": []}
```
And this gets reported as "problem" by the sensor. The sensor should check if the current array contains an error.

oischinger committed:
Move ActiveError back to climate component

The data returned by ActiveError is more complex.
It takes further investigation on how to interpret it a s a binary sensor.
Therefore it is moved back as an attribute of the climate component

oischinger commented:
@crazyfx1 For me the json seems to be empty.
I moved the ActiveError back as an attribute of the climate component for now.
I'll have to create some errors on my heater  to see how the data changes. Let's do this in a later PR.

crazyfx1 commented:
@oischinger With the current implementation I have the issue of running in the API limit every day. Do you see the same issue?

oischinger commented:
Yes, Viessmann must have changed something recently. I'm seeing this behavior since a few days but did not really find a pattern yet

crazyfx1 commented:
@oischinger It has something to do with our implementation in this PR. As soon as I rollback to the current version it works. But I can't really figure out whats the problem.

crazyfx1 commented:
I think I've found the problem and create a PR on the PyViCare repository. We should wait with the merge of this HA PR until the library is updated.

oischinger commented:
Nice catch. I wonder why this wasn't a problem with the water heater. But maybe it was but we just didn't notice.

crazyfx1 committed:
Update PyViCare library

crazyfx1 commented:
Seems like it's fixed.

oischinger commented:
Looks good to me, too.
Seems like we need an external review since we both committed?

oischinger reviewed (approved):
API limits have not been exceeded within the last week,

crazyfx1 commented:
We missed another release to get this in. 
I don't know the process to get this reviewed. This is one of the oldest PR on this repo now.

oischinger commented:
@MartinHjelmare I believe we need an external review for this since we both contributed. Could you please have a look?

Thanks alot!

MartinHjelmare commented on homeassistant/components/vicare/binary_sensor.py:
```diff
@@ -0,0 +1,123 @@
+    sensors = SENSORS_GENERIC
+
+    if heating_type != HeatingType.generic:
+        sensors.extend(SENSORS_BY_HEATINGTYPE[heating_type])
```
We're not allowed to modify the constant `sensors`. Either copy the constant or split the iteration.

MartinHjelmare replied:
I think it's simpler if just iterate all sensor types and use `continue` in a check.
```py
SENSORS_BY_HEATINGTYPE = {
    HeatingType.generic: SENSOR_CIRCULATION_PUMP_ACTIVE,
    HeatingType.gas: SENSOR_BURNER_ACTIVE,
    HeatingType.heatpump: SENSOR_COMPRESSOR_ACTIVE,
}


def setup_platform(hass, config, add_entities, discovery_info=None):
    """Create the ViCare sensor devices."""
    if discovery_info is None:
        return

    vicare_api = hass.data[VICARE_DOMAIN][VICARE_API]
    heating_type = hass.data[VICARE_DOMAIN][VICARE_HEATING_TYPE]
    name = hass.data[VICARE_DOMAIN][VICARE_NAME]

	for sensor in SENSOR_TYPES:
        sensor_type = SENSORS_BY_HEATINGTYPE[heating_type]
	    if sensor != sensor_type and sensor != SENSOR_CIRCULATION_PUMP_ACTIVE:
	        continue
	    
	    entity = ViCareBinarySensor(name, vicare_api, sensor)
	    entities.append(entity)

	add_entities(entities)
```

crazyfx1 replied:
I've decided to go for the copy method. I think it looks cleaner if you consider that you need to repeat all generic sensors in the loop in sensor.py.

MartinHjelmare commented on homeassistant/components/vicare/binary_sensor.py:
```diff
@@ -0,0 +1,123 @@
+class ViCareBinarySensor(BinarySensorDevice):
+    """Representation of a ViCare sensor."""
+
+    def __init__(self, name, api, sensortype):
```
Please separate words with underscore.
```suggestion
    def __init__(self, name, api, sensor_type):
```

MartinHjelmare commented on homeassistant/components/vicare/binary_sensor.py:
```diff
@@ -0,0 +1,123 @@
+
+    def __init__(self, name, api, sensortype):
+        """Initialize the sensor."""
+        self._sensor = SENSOR_TYPES[sensortype]
+        self._name = f"{name} {self._sensor[CONF_NAME]}"
+        self._api = api
+        self._sensortype = sensortype
```
```suggestion
        self._sensor = SENSOR_TYPES[sensor_type]
        self._name = f"{name} {self._sensor[CONF_NAME]}"
        self._api = api
        self._sensor_type = sensor_type
```

MartinHjelmare commented on homeassistant/components/vicare/binary_sensor.py:
```diff
@@ -0,0 +1,123 @@
+    @property
+    def unique_id(self):
+        """Return a unique ID."""
+        return f"{self._api.service.id}-{self._sensortype}"
```
```suggestion
        return f"{self._api.service.id}-{self._sensor_type}"
```

MartinHjelmare commented on homeassistant/components/vicare/binary_sensor.py:
```diff
@@ -0,0 +1,123 @@
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        if self._state is not None and self._state != PYVICARE_ERROR:
+            return True
+        return False
```
```suggestion
        return self._state is not None and self._state != PYVICARE_ERROR
```

MartinHjelmare commented on homeassistant/components/vicare/sensor.py:
```diff
@@ -0,0 +1,265 @@
+    sensors = SENSORS_GENERIC
+
+    if heating_type != HeatingType.generic:
+        sensors.extend(SENSORS_BY_HEATINGTYPE[heating_type])
```
See comment above.

MartinHjelmare commented on homeassistant/components/vicare/sensor.py:
```diff
@@ -0,0 +1,265 @@
+class ViCareSensor(Entity):
+    """Representation of a ViCare sensor."""
+
+    def __init__(self, name, api, sensortype):
+        """Initialize the sensor."""
+        self._sensor = SENSOR_TYPES[sensortype]
+        self._name = f"{name} {self._sensor[CONF_NAME]}"
+        self._api = api
+        self._sensortype = sensortype
```
```suggestion
    def __init__(self, name, api, sensor_type):
        """Initialize the sensor."""
        self._sensor = SENSOR_TYPES[sensor_type]
        self._name = f"{name} {self._sensor[CONF_NAME]}"
        self._api = api
        self._sensor_type = sensor_type
```

MartinHjelmare commented on homeassistant/components/vicare/sensor.py:
```diff
@@ -0,0 +1,265 @@
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        if self._state is not None and self._state != PYVICARE_ERROR:
+            return True
+        return False
```
```suggestion
        return self._state is not None and self._state != PYVICARE_ERROR
```

MartinHjelmare commented on homeassistant/components/vicare/sensor.py:
```diff
@@ -0,0 +1,265 @@
+    @property
+    def unique_id(self):
+        """Return a unique ID."""
+        return f"{self._api.service.id}-{self._sensortype}"
```
```suggestion
        return f"{self._api.service.id}-{self._sensor_type}"
```

MartinHjelmare commented on homeassistant/components/vicare/sensor.py:
```diff
@@ -0,0 +1,265 @@
+SENSOR_TYPES = {
+    SENSOR_OUTSIDE_TEMPERATURE: {
+        CONF_NAME: "Outside Temperature",
+        CONF_ICON: "mdi:thermometer",
```
Please set the temperature device_class and remove the custom icon.

Please review all sensor types and set the appropriate device class when possible.

Here are the available device classes for sensors:
https://developers.home-assistant.io/docs/core/entity/sensor#available-device-classes

crazyfx1 replied:
Thanks for the suggestion, I've added the device classes, I hope i got them right. 
For example it's not clear to me if a energy meter measured in kWh is also device class power, since it's not explicitly mentioned in the documentation ;)

MartinHjelmare replied:
No, those are separate measurements.

MartinHjelmare reviewed (changes requested):
Looks good! Some comments.

crazyfx1 committed:
PR changes

MartinHjelmare commented on homeassistant/components/vicare/sensor.py:
```diff
@@ -0,0 +1,289 @@
+        CONF_ICON: None,
+        CONF_UNIT_OF_MEASUREMENT: ENERGY_KILO_WATT_HOUR,
+        CONF_GETTER: lambda api: api.getGasConsumptionDomesticHotWaterToday(),
+        CONF_DEVICE_CLASS: DEVICE_CLASS_POWER,
```
Energy measurements with unit kWh should not have device class power.

crazyfx1 replied:
Ok thanks, I've changed it back :)

crazyfx1 committed:
PR changes 2

MartinHjelmare reviewed (approved):
Looks good!
